File: packages/swift-client/Sources/FirebaseAuth/BridgeAuthOperations.swift
import Foundation
import PyricFirestore

extension PyricBridgeClient {

    public func authSignInEmail(email: String, password: String, tenantId: String? = nil) async throws -> AnySendable {
        var params: [String: AnySendable] = [
            "email": .string(email),
            "password": .string(password)
        ]
        if let tenantId {
            params["tenantId"] = .string(tenantId)
        }
        return try await op(
            method: "auth.signInEmail",
            params: params
        )
    }

    public func authCreateUser(email: String, password: String, tenantId: String? = nil) async throws -> AnySendable {
        var params: [String: AnySendable] = [
            "email": .string(email),
            "password": .string(password)
        ]
        if let tenantId {
            params["tenantId"] = .string(tenantId)
        }
        return try await op(
            method: "auth.createUser",
            params: params
        )
    }

    public func authSignInAnonymously(tenantId: String? = nil) async throws -> AnySendable {
        var params: [String: AnySendable] = [:]
        if let tenantId {
            params["tenantId"] = .string(tenantId)
        }
        return try await op(
            method: "auth.signInAnonymously",
            params: params
        )
    }

    public func authSignOut() async throws {
        _ = try await op(
            method: "auth.signOut",
            params: [:]
        )
    }

    public func authGetIdToken(forcingRefresh: Bool = false) async throws -> String {
        let res = try await op(
            method: "auth.getIdToken",
            params: ["forceRefresh": .bool(forcingRefresh)]
        )
        guard let token = res.stringValue else {
            throw PyricBridgeError.internalError("Invalid token response from auth.getIdToken")
        }
        return token
    }

    public func authGetIdTokenResult(forcingRefresh: Bool = false) async throws -> AnySendable {
        try await op(
            method: "auth.getIdTokenResult",
            params: ["forceRefresh": .bool(forcingRefresh)]
        )
    }

    public func authGetCurrentUser() async throws -> AnySendable {
        try await op(
            method: "auth.getCurrentUser",
            params: [:]
        )
    }

    public func authUpdateProfile(displayName: String?, photoURL: String?) async throws -> AnySendable {
        var params: [String: AnySendable] = [:]
        if let displayName {
            params["displayName"] = .string(displayName)
        }
        if let photoURL {
            params["photoURL"] = .string(photoURL)
        }
        return try await op(
            method: "auth.updateProfile",
            params: params
        )
    }

    public func authRestorePortSession(uid: String, tenantId: String? = nil) async throws -> AnySendable {
        var params: [String: AnySendable] = ["uid": .string(uid)]
        if let tenantId {
            params["tenantId"] = .string(tenantId)
        }
        return try await op(
            method: "auth.restorePortSession",
            params: params
        )
    }
}

File: packages/flutter-client/lib/src/transport/bridge_auth_operations.dart
import 'bridge_client.dart';

/// Extension methods adding Auth-specific RPC conveniences to [PyricBridgeClient].
extension PyricBridgeAuthOps on PyricBridgeClient {
  /// Signs in a user with email and password.
  Future<dynamic> authSignInEmail(
    String email,
    String password, {
    String? tenantId,
  }) {
    return op('auth.signInEmail', {
      'email': email,
      'password': password,
      if (tenantId != null) 'tenantId': tenantId,
    });
  }

  /// Creates a new user with email and password.
  Future<dynamic> authCreateUser(
    String email,
    String password, {
    String? tenantId,
  }) {
    return op('auth.createUser', {
      'email': email,
      'password': password,
      if (tenantId != null) 'tenantId': tenantId,
    });
  }

  /// Signs in anonymously.
  Future<dynamic> authSignInAnonymously({String? tenantId}) {
    return op('auth.signInAnonymously', {
      if (tenantId != null) 'tenantId': tenantId,
    });
  }

  /// Signs out the current user session.
  Future<dynamic> authSignOut() {
    return op('auth.signOut', {});
  }

  /// Fetches the ID token for the current user.
  Future<String?> authGetIdToken({bool forceRefresh = false}) async {
    final res = await op('auth.getIdToken', {'forceRefresh': forceRefresh});
    return res as String?;
  }

  /// Fetches structured ID token results with claims and timestamps.
  Future<Map<String, dynamic>> authGetIdTokenResult({
    bool forceRefresh = false,
  }) async {
    final res = await op(
      'auth.getIdTokenResult',
      {'forceRefresh': forceRefresh},
    );
    return Map<String, dynamic>.from(res as Map);
  }

  /// Fetches the current user profile from the worker.
  Future<dynamic> authGetCurrentUser() {
    return op('auth.getCurrentUser', {});
  }

  /// Updates displayName and/or photoURL on the user profile.
  Future<dynamic> authUpdateProfile({String? displayName, String? photoURL}) {
    return op('auth.updateProfile', {
      if (displayName != null) 'displayName': displayName,
      if (photoURL != null) 'photoURL': photoURL,
    });
  }

  /// Updates email address on the user profile.
  Future<dynamic> authUpdateEmail(String newEmail) {
    return op('auth.updateEmail', {'newEmail': newEmail});
  }

  /// Updates password on the user profile.
  Future<dynamic> authUpdatePassword(String newPassword) {
    return op('auth.updatePassword', {'newPassword': newPassword});
  }

  /// Restores a previously persisted port session by UID.
  Future<dynamic> authRestorePortSession(String uid, {String? tenantId}) {
    return op('auth.restorePortSession', {
      'uid': uid,
      if (tenantId != null) 'tenantId': tenantId,
    });
  }

  /// Sets session persistence mode (e.g. 'LOCAL', 'SESSION', 'NONE').
  Future<void> authSetPersistence(String mode) async {
    await op('auth.setPersistence', {'mode': mode});
  }

  /// Lists all sandbox users via the auth.listUsers RPC.
  Future<List<Map<String, dynamic>>> authListUsers() async {
    final res = await op('auth.listUsers', {});
    if (res is List) {
      return res.map((item) => Map<String, dynamic>.from(item as Map)).toList();
    }
    return [];
  }
}
